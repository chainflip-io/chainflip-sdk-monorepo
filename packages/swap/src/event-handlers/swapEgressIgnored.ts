import { z } from 'zod';
import { u64, chainflipAssetEnum, u128, hexString } from '@/shared/parsers';
import { getStateChainError } from './common';
import { EventHandlerArgs } from './index';

const swapEgressIgnoredArgs = z.object({
  asset: chainflipAssetEnum,
  amount: u128,
  swapId: u64,
  reason: z.object({
    __kind: z.literal('Module'),
    value: z.object({
      error: hexString,
      index: z.number(),
    }),
  }),
});

export type SwapDepositAddressReadyEvent = z.input<
  typeof swapEgressIgnoredArgs
>;

export default async function swapEgressIgnored({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const { amount, swapId, reason } = swapEgressIgnoredArgs.parse(event.args);

  const [failure, swap] = await Promise.all([
    getStateChainError(prisma, block, reason.value),
    prisma.swap.findUniqueOrThrow({ where: { nativeId: swapId } }),
  ]);

  await prisma.ignoredEgress.create({
    data: {
      swapId: swap.id,
      ignoredAt: new Date(block.timestamp),
      ignoredBlockIndex: `${block.height}-${event.indexInBlock}`,
      amount: amount.toString(),
      stateChainErrorId: failure.id,
    },
  });
}
