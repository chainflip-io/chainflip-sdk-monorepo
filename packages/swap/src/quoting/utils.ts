import assert from 'assert';
import BigNumber from 'bignumber.js';
import { filter, Observable, Subscription } from 'rxjs';
import { Side, type SwapInput } from '@/amm-addon';
import { InternalAsset, getAssetAndChain } from '@/shared/enums';
import { MarketMakerQuote, Leg as MarketMakerLeg } from './schemas';
import env from '../config/env';
import { getAssetPrice } from '../pricing';
import { assertUnreachable } from '../utils/function';
import { SwapRateArgs } from '../utils/statechain';

export type QuoteType = 'pool' | 'market_maker';

export const collectMakerQuotes = (
  requestId: string,
  expectedQuotes: number,
  quotes$: Observable<{ marketMaker: string; quote: MarketMakerQuote }>,
): Promise<MarketMakerQuote[]> => {
  if (expectedQuotes === 0) return Promise.resolve([]);

  const clientsReceivedQuotes = new Map<string, MarketMakerQuote>();

  return new Promise((resolve) => {
    let sub: Subscription;

    const complete = () => {
      sub.unsubscribe();
      resolve([...clientsReceivedQuotes.values()]);
    };

    sub = quotes$
      .pipe(filter(({ quote }) => quote.request_id === requestId))
      .subscribe(({ marketMaker, quote }) => {
        clientsReceivedQuotes.set(marketMaker, quote);
        if (clientsReceivedQuotes.size === expectedQuotes) complete();
      });

    setTimeout(complete, env.QUOTE_TIMEOUT);
  });
};

export const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
  const price = await getAssetPrice(asset);

  if (typeof price !== 'number') return null;

  return BigInt(new BigNumber(amount).times(price).toFixed(0));
};

type BaseAsset = Exclude<InternalAsset, 'Usdc'>;

export class Leg {
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint): Leg;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null): Leg | null;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null) {
    if (amount === null) return null;
    assert(baseAsset !== quoteAsset, 'baseAsset and quoteAsset must be different');
    assert(baseAsset === 'Usdc' || quoteAsset === 'Usdc', 'one of the assets must be USDC');
    return new Leg(baseAsset, quoteAsset, amount);
  }

  private constructor(
    private readonly baseAsset: InternalAsset,
    private readonly quoteAsset: InternalAsset,
    public amount: bigint,
  ) {}

  toPoolJSON(): SwapRateArgs {
    return {
      amount: this.amount,
      srcAsset: this.baseAsset,
      destAsset: this.quoteAsset,
    };
  }

  private getSide(): 'BUY' | 'SELL' {
    return this.quoteAsset !== 'Usdc' ? 'BUY' : 'SELL';
  }

  toMarketMakerJSON(): MarketMakerLeg {
    const side = this.getSide();
    let normalizedBaseAsset: BaseAsset;
    const normalizedQuoteAsset = 'Usdc';

    if (this.quoteAsset !== 'Usdc') {
      normalizedBaseAsset = this.quoteAsset;
    } else if (this.baseAsset !== 'Usdc') {
      normalizedBaseAsset = this.baseAsset;
    } else {
      return assertUnreachable('invalid leg');
    }

    return {
      base_asset: getAssetAndChain(normalizedBaseAsset),
      quote_asset: getAssetAndChain(normalizedQuoteAsset),
      amount: this.amount.toString(),
      side,
    };
  }

  toSwapInput(): SwapInput {
    return {
      side: this.getSide() === 'BUY' ? Side.Buy : Side.Sell,
      amount: this.amount,
      limitOrders: [],
    };
  }
}
