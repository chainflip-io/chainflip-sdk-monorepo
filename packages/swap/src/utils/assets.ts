import { SupportedAsset, supportedAsset } from '@/shared/enums';

export const isSupportedAsset = (value: string): value is SupportedAsset =>
  supportedAsset.safeParse(value).success;

export function assertSupportedAsset(
  value: string,
): asserts value is SupportedAsset {
  if (!isSupportedAsset(value)) {
    throw new Error(`received invalid asset "${value}"`);
  }
}

export const decimalPlaces: Record<SupportedAsset, number> = {
  DOT: 10,
  ETH: 18,
  FLIP: 18,
  USDC: 6,
  BTC: 8,
};
