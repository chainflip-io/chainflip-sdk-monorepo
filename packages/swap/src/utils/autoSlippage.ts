import BigNumber from 'bignumber.js';
import { assetConstants, chainConstants } from '@/shared/enums';
import { getDeployedLiquidity, getUndeployedLiquidity } from './pools';
import { getRequiredBlockConfirmations } from './rpc';
import { InternalAsset } from '../client';

const getLiquidityAutoSlippage = async (
  srcAsset: InternalAsset,
  destAsset: InternalAsset,
  amount: bigint,
) => {
  const deployedLiquidity = await getDeployedLiquidity(srcAsset, destAsset);
  const liquidityRatio = new BigNumber(amount.toString()).div(deployedLiquidity.toString());

  if (liquidityRatio.isGreaterThan(0.2)) {
    return liquidityRatio.multipliedBy(2);
  }
  return new BigNumber(0);
};

const getDepositTimeAutoSlippage = async (srcAsset: InternalAsset, isBoosted: boolean) => {
  const { chain } = assetConstants[srcAsset];
  const { blockTimeSeconds } = chainConstants[chain];
  const blockConfirmations = isBoosted ? 1 : (await getRequiredBlockConfirmations(srcAsset)) ?? 0;
  const depositTimeMinutes = (blockConfirmations * blockTimeSeconds) / 60;

  if (['Solana', 'Arbitrum'].includes(chain)) {
    return depositTimeMinutes * 0.01;
  }
  if (['Ethereum', 'Polkadot'].includes(chain)) {
    return depositTimeMinutes * 0.05;
  }
  return depositTimeMinutes * 0.1;
};

const getUndeployedLiquidityAutoSlippage = async (asset: InternalAsset, amount: bigint) => {
  const undeployedLiquidity = await getUndeployedLiquidity(asset);
  if (undeployedLiquidity >= amount) {
    return new BigNumber(-0.5);
  }

  const undeployedRatio = new BigNumber(undeployedLiquidity.toString()).div(amount.toString());
  if (undeployedRatio.gt(0.5)) {
    return new BigNumber(-undeployedRatio * 0.1);
  }

  return new BigNumber(0);
};

export const calculateRecommendedSlippage = async ({
  srcAsset,
  destAsset,
  intermediateAmount,
  egressAmount,
  boostFeeBps,
  dcaChunks,
}: {
  srcAsset: InternalAsset;
  destAsset: InternalAsset;
  intermediateAmount?: bigint;
  egressAmount: bigint;
  boostFeeBps?: number;
  dcaChunks: number;
}) => {
  const minSlippage = 0.1;
  const baseSlippage = 1;

  const assetTo = srcAsset === 'Usdc' || destAsset === 'Usdc' ? destAsset : 'Usdc';

  const [deployedLiquiditySlippage, timeSlippage, undeployedLiquiditySlippage] = await Promise.all([
    await getLiquidityAutoSlippage(
      srcAsset,
      assetTo,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
    await getDepositTimeAutoSlippage(srcAsset, Boolean(boostFeeBps && boostFeeBps > 0)),
    await getUndeployedLiquidityAutoSlippage(
      intermediateAmount ? 'Usdc' : destAsset,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
  ]);

  const calculatedSlippage = deployedLiquiditySlippage
    .plus(timeSlippage)
    .plus(undeployedLiquiditySlippage)
    .plus(baseSlippage);

  return Math.max(minSlippage, Math.round(calculatedSlippage.toNumber() * 100) / 100);
};
