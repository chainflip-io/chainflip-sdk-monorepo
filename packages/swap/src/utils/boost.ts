import { ONE_IN_PIP, bigintMin, getPipAmountFromAmount } from '@/shared/functions';
import { BoostPoolsDepth } from '@/shared/rpc';

export const getBoostFeeBpsForAmount = async ({
  amount,
  assetBoostPoolsDepth,
}: {
  amount: bigint;
  assetBoostPoolsDepth: BoostPoolsDepth;
}): Promise<number | undefined> => {
  assetBoostPoolsDepth.sort((a, b) => (a.tier < b.tier ? -1 : 1));

  let _amount = amount;
  let feeAmount = 0n;

  for (const poolDepth of assetBoostPoolsDepth) {
    const poolAvailableAmount = poolDepth.availableAmount;

    const amountToBeUsedFromPool = bigintMin(_amount, poolAvailableAmount);
    feeAmount += getPipAmountFromAmount(amountToBeUsedFromPool, poolDepth.tier);
    _amount -= amountToBeUsedFromPool;
  }

  // Not enough liquidity in the boost pools
  if (_amount > 0) return undefined;

  return Number((feeAmount * BigInt(ONE_IN_PIP)) / BigInt(amount));
};
