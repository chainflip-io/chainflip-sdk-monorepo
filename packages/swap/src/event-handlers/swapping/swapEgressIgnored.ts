import { swappingSwapEgressIgnored as schema160 } from '@chainflip/processor/160/swapping/swapEgressIgnored';
import { swappingSwapEgressIgnored as schema190 } from '@chainflip/processor/190/swapping/swapEgressIgnored';
import { z } from 'zod';
import { getStateChainError } from '../common.js';
import { EventHandlerArgs } from '../index.js';

const swapEgressIgnoredArgs = z.union([schema190, schema160]);

export type SwapEgressIgnoredArgs = z.input<typeof swapEgressIgnoredArgs>;

export default async function swapEgressIgnored({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const { amount, swapRequestId, reason } = swapEgressIgnoredArgs.parse(event.args);

  const failure = await (reason.__kind === 'Module'
    ? getStateChainError(prisma, block, reason.value)
    : prisma.stateChainError.upsert({
        where: {
          specVersion_palletIndex_errorIndex: { specVersion: 0, palletIndex: 0, errorIndex: 0 },
        },
        create: {
          specVersion: 0,
          palletIndex: 0,
          errorIndex: 0,
          docs: 'Unknown error',
          name: 'Unknown error',
        },
        update: {},
      }));

  await prisma.ignoredEgress.create({
    data: {
      ignoredAt: new Date(block.timestamp),
      ignoredBlockIndex: `${block.height}-${event.indexInBlock}`,
      amount: amount.toString(),
      stateChainError: { connect: { id: failure.id } },
      swapRequest: { connect: { nativeId: swapRequestId } },
      type: 'SWAP',
    },
  });
}
