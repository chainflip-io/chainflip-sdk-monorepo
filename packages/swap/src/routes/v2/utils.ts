import { getAssetAndChain } from '@/shared/enums';
import { assertUnreachable } from '@/shared/functions';
import { StateV2 } from './swap';
import prisma, {
  Prisma,
  Swap,
  Egress,
  Broadcast,
  SwapFee,
  IgnoredEgress,
  FailedSwap,
} from '../../client';
import { getPendingBroadcast } from '../../ingress-egress-tracking';
import { failedSwapMessage } from '../../utils/swap';
import { FailureMode } from '../swap';

export const depositChannelInclude = {
  failedBoosts: true,
  failedSwaps: true,
  affiliates: {
    select: {
      account: true,
      commissionBps: true,
    },
  },
} as const;

export const swapRequestInclude = {
  swaps: { include: { fees: true }, orderBy: { nativeId: 'desc' } },
  egress: { include: { broadcast: true } },
  refundEgress: { include: { broadcast: true } },
  fees: true,
  ignoredEgresses: true,
  swapDepositChannel: {
    include: depositChannelInclude,
  },
} as const;

export const getSwapFields = (swap: Swap & { fees: SwapFee[] }) => ({
  swapInputAmount: swap.swapInputAmount.toString(),
  swapOutputAmount: swap.swapOutputAmount?.toString(),
  scheduledAt: swap.swapScheduledAt.valueOf(),
  scheduledBlockIndex: swap.swapScheduledBlockIndex ?? undefined,
  executedAt: swap.swapExecutedAt?.valueOf(),
  executedBlockIndex: swap.swapExecutedBlockIndex ?? undefined,
  retryCount: swap.retryCount,
  latestSwapRescheduledAt: swap.latestSwapRescheduledAt?.valueOf(),
  latestSwapRescheduledBlockIndex: swap.latestSwapRescheduledBlockIndex ?? undefined,
  fees: swap.fees.map((fee) => ({
    type: fee.type,
    ...getAssetAndChain(fee.asset),
    amount: fee.amount.toString(),
  })),
});

export const getDepositIgnoredFailedState = (failedSwap: FailedSwap) => ({
  failedAt: failedSwap.failedAt.valueOf(),
  failedAtBlockIndex: failedSwap.failedBlockIndex,
  mode: FailureMode.IngressIgnored,
  reason: {
    name: failedSwap.reason,
    message: failedSwapMessage[failedSwap.reason],
  },
});

export const getEgressFailureState = async (
  ignoredEgress: IgnoredEgress | undefined,
  broadcast: Broadcast | null | undefined,
  type: IgnoredEgress['type'] | undefined,
) => {
  const abortedBroadcast = broadcast?.abortedAt?.valueOf();
  const failedAt = abortedBroadcast ?? ignoredEgress?.ignoredAt?.valueOf();
  const failedAtBlockIndex = broadcast?.abortedBlockIndex ?? ignoredEgress?.ignoredBlockIndex;

  let error;
  let failureMode;

  if (failedAt && failedAtBlockIndex) {
    error = {
      name: 'Unknown',
      message: 'An unknown error occurred',
    };
    if (ignoredEgress) {
      switch (ignoredEgress.type) {
        case 'REFUND':
          failureMode = FailureMode.RefundEgressIgnored;
          break;
        case 'SWAP':
          failureMode = FailureMode.EgressIgnored;
          break;
        default:
          assertUnreachable(ignoredEgress.type);
      }
      const [stateChainError] = await Promise.all([
        prisma.stateChainError.findUniqueOrThrow({
          where: { id: ignoredEgress.stateChainErrorId },
        }),
      ]);
      error = {
        name: stateChainError.name,
        message: stateChainError.docs,
      };
    } else if (abortedBroadcast) {
      const message =
        type === 'SWAP' ? 'The swap broadcast was aborted' : 'The refund broadcast was aborted';
      failureMode = FailureMode.BroadcastAborted;
      error = {
        name: 'BroadcastAborted',
        message,
      };
    }
    return {
      failedAt,
      failedAtBlockIndex,
      mode: failureMode,
      reason: error,
    };
  }
  return undefined;
};

export const getEgressStatusFields = async (
  egress: Egress | null | undefined,
  broadcast: Broadcast | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  type: IgnoredEgress['type'] | undefined,
  egressTrackerTxRef: string | null | undefined,
) => {
  const ignoredEgress = ignoredEgresses?.find((e) => e.type === type);
  const failureState = await getEgressFailureState(ignoredEgress, broadcast, type);
  return {
    ...(egress && {
      outputAmount: egress.amount?.toFixed(),
      scheduledAt: egress.scheduledAt?.valueOf(),
      scheduledBlockIndex: egress.scheduledBlockIndex ?? undefined,
    }),
    ...(broadcast && {
      sentAt: broadcast?.succeededAt?.valueOf(),
      sentAtBlockIndex: broadcast?.succeededBlockIndex ?? undefined,
      sentTxRef: broadcast?.transactionRef ?? egressTrackerTxRef,
    }),
    ...(failureState && { failure: failureState }),
    ...(ignoredEgress && { ignoredAmount: ignoredEgress.amount?.toFixed() }),
  };
};

export const getSwapState = async (
  failedSwap: FailedSwap | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  swapRequest:
    | Prisma.SwapRequestGetPayload<{
        include: typeof swapRequestInclude;
      }>
    | undefined
    | null,
) => {
  let state: StateV2 | undefined;
  let swapEgressTrackerTxRef: string | null | undefined;
  let refundEgressTrackerTxRef: string | null | undefined;
  const swapEgress = swapRequest?.egress;
  const refundEgress = swapRequest?.refundEgress;
  const egress = swapEgress ?? refundEgress;

  if (
    failedSwap ||
    (ignoredEgresses && ignoredEgresses.length > 0) ||
    egress?.broadcast?.abortedAt
  ) {
    state = StateV2.Failed;
  } else if (egress?.broadcast?.succeededAt) {
    state = StateV2.Completed;
  } else if (egress?.broadcast) {
    if (swapEgress?.broadcast) {
      state = StateV2.Sending;
      const pendingSwapBroadcast = await getPendingBroadcast(swapEgress.broadcast);
      if (pendingSwapBroadcast) {
        state = StateV2.Sent;
        swapEgressTrackerTxRef = pendingSwapBroadcast.tx_ref;
      }
    }
    if (refundEgress?.broadcast) {
      const pendingRefundBroadcast = await getPendingBroadcast(refundEgress.broadcast);
      if (pendingRefundBroadcast) {
        state = StateV2.Sent;
        refundEgressTrackerTxRef = pendingRefundBroadcast.tx_ref;
      }
    }
  } else if (egress) {
    state = StateV2.Sending;
  } else if (swapRequest?.swaps.some((s) => s.swapScheduledAt)) {
    state = StateV2.Swapping;
  } else {
    state = StateV2.Receiving;
  }

  return {
    state,
    swapEgressTrackerTxRef,
    refundEgressTrackerTxRef,
  };
};
