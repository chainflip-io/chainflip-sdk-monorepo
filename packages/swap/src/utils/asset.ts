import { ChainflipAsset } from '@chainflip/utils/chainflip';

export const isAssetHub = (asset: ChainflipAsset): asset is 'HubDot' | 'HubUsdc' | 'HubUsdt' =>
  asset === 'HubDot' || asset === 'HubUsdc' || asset === 'HubUsdt';
