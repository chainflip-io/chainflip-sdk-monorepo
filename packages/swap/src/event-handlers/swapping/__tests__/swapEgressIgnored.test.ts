import { describe, it, expect, beforeEach, vi } from 'vitest';
import prisma from '../../../client';
import metadataMock from '../../__tests__/metadata.json';
import { DOT_ADDRESS, swapEgressIgnoredMock } from '../../__tests__/utils';
import swapEgressIgnored from '../swapEgressIgnored';

vi.mock('@/shared/rpc', async (importOriginal) => {
  const original = (await importOriginal()) as object;
  return {
    ...original,
    getMetadata: vi.fn().mockResolvedValue(metadataMock.result),
  };
});
const { event, block } = swapEgressIgnoredMock;

describe(swapEgressIgnored, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap", "SwapRequest", "Egress", "IgnoredEgress", "StateChainError" CASCADE`;
  });

  it('adds an IgnoredEgress record for a egress ignored event', async () => {
    const { swapRequestId } = event.args;

    await prisma.swapRequest.create({
      data: {
        nativeId: BigInt(swapRequestId),
        srcAsset: 'Btc',
        destAsset: 'Eth',
        destAddress: DOT_ADDRESS,
        depositAmount: '100000000',
        swapInputAmount: '100000000',
        depositFinalisedAt: new Date('2024-08-06T00:00:00.000Z'),
        depositFinalisedBlockIndex: '1-1',
        originType: 'VAULT',
        requestType: 'LEGACY_SWAP',
        swapRequestedAt: new Date('2024-08-06T00:00:00.000Z'),
        swapRequestedBlockIndex: '92-398',
        totalBrokerCommissionBps: 0,
      },
    });

    await swapEgressIgnored({ block, event, prisma });

    expect(
      await prisma.swapRequest.findUniqueOrThrow({
        where: { nativeId: BigInt(swapRequestId) },
        include: { ignoredEgresses: true },
      }),
    ).toMatchSnapshot({
      id: expect.any(BigInt),
      ignoredEgresses: [
        {
          id: expect.any(BigInt),
          swapRequestId: expect.any(BigInt),
          createdAt: expect.any(Date),
          stateChainErrorId: expect.any(Number),
        },
      ],
    });
  });
});
