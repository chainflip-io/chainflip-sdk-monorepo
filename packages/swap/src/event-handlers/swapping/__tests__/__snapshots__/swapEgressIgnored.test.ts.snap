// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`swapEgressIgnored > adds an IgnoredEgress record for a egress ignored event 1`] = `
{
  "ccmGasBudget": null,
  "ccmMessage": null,
  "completedAt": null,
  "completedBlockIndex": null,
  "dcaChunkIntervalBlocks": null,
  "dcaNumberOfChunks": null,
  "depositAmount": "100000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositFinalisedAt": 2024-08-06T00:00:00.000Z,
  "depositFinalisedBlockIndex": "1-1",
  "depositTransactionRef": null,
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "Eth",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "fallbackEgressId": null,
  "fokMaxOraclePriceSlippageBps": null,
  "fokMinPriceX128": null,
  "fokRefundAddress": null,
  "fokRetryDurationBlocks": null,
  "id": Any<BigInt>,
  "ignoredEgresses": [
    {
      "amount": "11",
      "createdAt": Any<Date>,
      "id": Any<BigInt>,
      "ignoredAt": 2022-12-06T14:31:45.000Z,
      "ignoredBlockIndex": "120-0",
      "stateChainErrorId": Any<Number>,
      "swapRequestId": Any<BigInt>,
      "type": "SWAP",
    },
  ],
  "maxBoostFeeBps": 0,
  "nativeId": 2n,
  "oraclePriceDeltaBps": null,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgressId": null,
  "requestType": "LEGACY_SWAP",
  "srcAddress": null,
  "srcAsset": "Btc",
  "swapDepositChannelId": null,
  "swapInputAmount": "100000000",
  "swapOutputAmount": null,
  "swapRequestedAt": 2024-08-06T00:00:00.000Z,
  "swapRequestedBlockIndex": "92-398",
  "totalBrokerCommissionBps": 0,
}
`;
